Build generic eval blocks incrementally

zygoloid opened the pull request:
Instead of building an eval block as a separate pass at the end of a generic, build the eval block incrementally.

The larger change here is that asking for the type or constant value of an instruction now always returns an unattached type or constant value, in order to preserve the behavior that we previously achieved by doing the rewrite to attached types and constant values at the end of handling the generic.

This also incidentally fixes some subtle issues where attached types and constant values would leak out into check and cause it to get confused about differences between attached and unattached values. Check should no longer see attached values except where it explicitly asks for them.

zygoloid committed:
Checkpoint progress.

zygoloid committed:
Progress

zygoloid committed:
A couple more fixes.

zygoloid committed:
Fix template handling.

zygoloid committed:
More template fixing.

zygoloid committed:
Handle redeclarations.

zygoloid committed:
Return unattached constants from constant_values().Get().

zygoloid committed:
cleanups and clang-format

zygoloid committed:
Cleanup.

zygoloid committed:
One last bug.

zygoloid committed:
clang-format

zygoloid committed:
Use `TypeInstId`.

zygoloid committed:
Another TODO.

zygoloid committed:
Typo.

zygoloid committed:
Clean up import handling: don't get attached types by default there
either.

zygoloid committed:
Build generics on the fly.

zygoloid committed:
clang-format

zygoloid commented:
Depends on #5328.

danakj commented:
Could we rebase this first as its dependency has merged, and it has conflicts?

zygoloid committed:
Autoupdate no_prelude tests.

Differences:
- Specifics appear in different orders.
- In one test we now correctly notice we can forward an initializing
  expression unchanged.

zygoloid committed:
More fixes.

danakj commented on toolchain/check/eval.cpp:
```diff
@@ -105,53 +128,33 @@ class EvalContext {
-  }
+    // While resolving a specific, map from previous instructions in the eval
+    // block into their evaluated values. These values won't be present on the
+    // specific itself yet, so `GetConstantValueInSpecific` won't be able to
```
```suggestion
    // specific itself yet, so `GetTypeOfInstInSpecific` won't be able to
```
right?

danakj commented on toolchain/check/subst.cpp:
```diff
@@ -256,9 +261,10 @@ static auto Rebuild(Context& context, Worklist& worklist, SemIR::InstId inst_id,
-                     : SemIR::TypeId::None;
+  auto type_id =
+      inst.type_id().has_value()
+          ? callbacks.RebuildType(SemIR::TypeInstId::UnsafeMake(worklist.Pop()))
```
I think it would be best to avoid the UnsafeMake here. We believe it should be a type but there's nothing here that guarantees that locally by construction, or by checks. And in the presence of bugs that we write in the future, going through an unsafe incorrectly makes a buggy situation into a much more confusing buggy situation.
```suggestion
          ? callbacks.RebuildType(context.types().GetAsTypeInstId(worklist.Pop()))
```

danakj commented on toolchain/check/member_access.cpp:
```diff
@@ -665,13 +665,13 @@ static auto GetAssociatedValueImpl(Context& context, SemIR::LocId loc_id,
   auto value_inst_id =
-      context.constant_values().GetConstantInstId(assoc_entity_inst_id);
+      context.constant_values().GetInstId(assoc_entity_const_id);
   CARBON_CHECK(value_inst_id.has_value());
```
This check probably makes more sense as checking `assoc_entity_const_id` has a (constant) value now? If it does GetInstId can't return None

zygoloid replied:
That's a good point. In fact, since we generally don't check that ID arguments are non-`None`, I'm inclined to remove this CHECK entirely.
```suggestion
```

danakj commented on toolchain/check/import_ref.cpp:
```diff
@@ -1132,27 +1134,27 @@ static auto GetLocalParamPatternsId(ImportContext& context,
 
-    auto binding =
-        context.import_insts().GetAs<SemIR::AnyBindingPattern>(binding_id);
+    auto binding_id = inst_id;
```
Before it was setting this to addr_pattern_id, now it may be the param_pattern->subpattern_id. Is that intended?

zygoloid replied:
(Same.)

danakj commented on toolchain/check/import_ref.cpp:
```diff
@@ -1132,27 +1134,27 @@ static auto GetLocalParamPatternsId(ImportContext& context,
-        context.import_insts().TryGetAs<SemIR::ValueParamPattern>(
-            param_pattern_id);
-    auto binding_id = addr_pattern_id;
+    auto param_pattern_id = inst_id;
```
Before it was setting this to addr_pattern_id, now it may be addr_inst->inner_id. Is that intended?

zygoloid replied:
Note the `if` on old lines 1142-1143. Unless I'm missing something, this was previously initialized to `addr_pattern_id` and then conditionally updated to `addr_inst->inner_id`; now we set it directly to the desired value.

danakj replied:
Ah, right ok good

danakj commented on toolchain/check/generic_region_stack.h:
```diff
@@ -6,68 +6,119 @@
+
+  // Adds an instruction to the list of instructions whose type or value depends
+  // on something in the current pending generic.
+  auto AddDependentInst(SemIR::InstId inst_id) -> void {
```
It would be nice if this and the free function in generic.h didn't share the exact same name, but I am not sure what to suggest, so if you have any ideas...

zygoloid replied:
Changed the free function to `AttachDependentInstToCurrentGeneric`.

danakj replied:
Thanks, it's nicer, and the new comment on it is good too :)

danakj commented on toolchain/check/generic.cpp:
```diff
@@ -25,21 +24,6 @@ namespace Carbon::Check {
-}
-
 // Adds an instruction `generic_inst_id` to the eval block for a generic region,
 // which is the current instruction block. The instruction `generic_inst_id` is
```
Does this comment need an update? We're not using the "current instruction block" anymore.

zygoloid replied:
Done. Updating this made me notice that this function doesn't really have a coherent interface any more: it takes a generic ID and a region, but always adds the instruction to the current pending generic's eval block. Changed it to always use the current pending generic's ID and region too.

danakj replied:
Nice, thanks!

danakj commented on toolchain/check/generic.cpp:
```diff
@@ -77,44 +53,43 @@ class RebuildGenericConstantInEvalBlockCallbacks : public SubstInstCallbacks {
-      // In order to reuse instructions from the generic as often as possible,
-      // keep this instruction as-is if it already has the desired symbolic
-      // constant value.
-      if (const_id != context_->constant_values().Get(result.value())) {
```
Was this conditional no longer doing anything? I'm having trouble understanding why in particular it goes away

zygoloid replied:
This condition stopped doing anything a long time ago. The instructions stored in `constants_in_generic_` are instructions in the eval block, and their constant values are always attached constants, whereas the instructions that we're substituting into are the instructions that define an unattached constant. So, in the old world, their constant values couldn't be the same; in the new world, the (unattached) constant values can be the same, but if so, it would be wrong to reuse the same instruction.

We *could* compare whether the attached constant values of the instructions are the same, but that check would never fire -- and even if it did, it'd result in the eval block containing an instruction that refers to an instruction in the body of the generic (not in the eval block), which seems undesirable.

(The original idea here had been that we would add the `InstId` of an instruction in the generic body directly to the eval block (rather than making a copy) if we could reuse that same instruction directly to compute the value of the constant in each specific. But we ended up not doing that, and this check is left over from that old approach.)

danakj replied:
Thanks, I think this all makes sense to me now :)

danakj commented on toolchain/check/generic.cpp:
```diff
@@ -312,83 +278,84 @@ static auto PopulateConstantsFromDeclaration(
+  //   // Has generic type and constant value, but no generic region.
+  //   fn A(T:! type).F() {}
+  //
+  // TODO: Use a different instruction kind for out-of-line definitions and
```
Can update this to say we want to copy the constant values from the generic decl?

zygoloid replied:
Done.

danakj commented on toolchain/check/generic.cpp:
```diff
@@ -312,83 +278,84 @@ static auto PopulateConstantsFromDeclaration(
+  // TODO: Use a different instruction kind for out-of-line definitions and
+  // CHECK this doesn't happen.
+  if (context.generic_region_stack().Empty()) {
+    if ((dep_kind & DependentInst::Template) != DependentInst::None) {
```
Could the comment explain why we put template-dependent insts into the current inst block, and only when the stack is empty?

zygoloid replied:
Looks like we don't need to handle this case any more. Converted to a CHECK.

danakj replied:
Is it because in a redecl we don't try to attach template-dependent instructions to the generic? Or that the generic stack wouldn't be empty if there's a template-dependent instruction? Or we never attach template-dependent instructions to a generic? Or something else?

zygoloid replied:
Added a comment to explain.

danakj commented on toolchain/check/generic.cpp:
```diff
@@ -312,83 +278,84 @@ static auto PopulateConstantsFromDeclaration(
+  if ((dep_kind & DependentInst::SymbolicType) != DependentInst::None) {
+    auto inst = context.insts().Get(inst_id);
+    auto type_id = AddGenericTypeToEvalBlock(
+        context, generic_id, region, context.insts().GetLocId(inst_id),
```
```suggestion
        context, generic_id, region, /*loc_id=*/inst_id,
```
We can pass `inst_id` as a location now

zygoloid replied:
I've tried that but it doesn't work:
```
FATAL failure at toolchain/sem_ir/absolute_node_id.cpp:81: Unexpected LocId: loc_inst44
```
so I've backed it out again. I guess this causes us to create an instruction whose location is another instruction, which is apparently disallowed but not enforced by the type system any more?

danakj replied:
Right, sorry, this breaks various things that consume LocId at the moment, and will be addressed by https://github.com/carbon-language/carbon-lang/pull/5355

danakj commented on toolchain/check/generic.cpp:
```diff
@@ -490,38 +503,147 @@ auto BuildGenericDecl(Context& context, SemIR::InstId decl_id)
+  if (auto [old_inst_id, new_inst_id] = FirstDifferenceBetweenEvalBlocks(
+          context, old_eval_block, new_eval_block);
+      new_inst_id.has_value()) {
+    context.TODO(new_inst_id,
```
Maybe CARBON_FATAL?

zygoloid replied:
I'm torn on this. This *should* be impossible, but the chain of logic under which this is impossible is quite tenuous and depends on there not existing anything in the language that could result in a difference without a token difference. This doesn't really feel like a precondition violation or similar thing for which CHECK would be appropriate, but also doesn't really feel like a diagnostic fits well because I'm not aware of any way to test it. While developing this PR, having the two locations here was useful. I suppose I could produce the TODOs and then a CARBON_FATAL instead of recovering from the error?

But as I say I'm torn. Happy to do whatever you prefer now I've given my thinking :) Let me know.

danakj replied:
I agree printing things can be helpful in the case of an error. Maybe a context.UNEXPECTED() would fit for cases like this a bit better than context.TODO(), but no need to do that now. Putting a FATAL after to TODOs isn't as helpful as might sound because then when running a file test it won't rewrite the file so you will have a hard time ever seeing the printed diagnostics. So it's okay to leave as is I think, but maybe we can think about changing the naming of TODO since that looks misleading. If you agree, maybe drop a TODO on the TODOs (^_^') about renaming them?

zygoloid replied:
Done, I added a big comment :)

For what it's worth, we do include the pending diagnostics with the backtrace in case of a crash, so TODO followed by a CARBON_FATAL should provide the extra information (though not in the test file).

danakj commented on toolchain/check/testdata/class/generic/member_out_of_line.carbon:
```diff
@@ -454,17 +456,19 @@ fn Generic(T:! ()).WrongType() {}
 // CHECK:STDOUT: specific @B(%T, %N.loc5_11.2) {}
 // CHECK:STDOUT:
-// CHECK:STDOUT: specific @A(%T.loc4_9.2) {}
+// CHECK:STDOUT: specific @B(<unexpected>.inst79.loc10_33, <unexpected>.inst80.loc10_33) {}
```
Something went wrong here

zygoloid replied:
Yeah, what's happening here is that we generate a new generic for the out of line declaration of `fn Class(T:! type).G`, including a new `T:! type` and a new non-canonical specific that holds the sugared name for the type argument `T`. And then we find it's a redeclaration, merge it with the previous declaraiton of `G`, and discard the new generic. But this leaves behind some junk in SemIR, such as this unreferenced specific.

A couple of options, then:

* When formatting SemIR, we could track which specifics are referenced, like we track which constants are referenced, and omit unreferenced specifics.
* When formatting SemIR, we could omit these non-canonical specifics entirely -- if a specific has no value blocks, we're not providing any information by printing it out.

I'm going with the second option. Sorry about the test churn! Let me know if you'd prefer something else.

danakj replied:
The latter sounds good, which I see you did. What do you think about turning that conditional check into a helper method on the specific? Is this something lowering should be checking then as well? It could share the same `HasValueBlock()` function or `WasNotResolved()` (or maybe something more prescriptive that the specific should be ignored) which can include a comment explaining what you've written here?

zygoloid replied:
Done. I'm not sure if lowering will want this, because it doesn't iterate over specifics looking for ones to lower, but it seems reasonable to factor out anyway.

danakj commented on toolchain/check/testdata/class/generic_method.carbon:
```diff
@@ -141,12 +141,14 @@ fn Class(T:! type).F[self: Self](n: T) {}
-// CHECK:STDOUT: specific @Class(@F.%T.loc13) {}
-// CHECK:STDOUT:
-// CHECK:STDOUT: specific @Class(%T.loc11_13.2) {}
+// CHECK:STDOUT: specific @Class(<unexpected>.inst66.loc16_28) {}
```
And here. I wonder if it's related to the `@Class` disappearing on L53? Or maybe DiscardGeneric?

danakj commented on toolchain/check/testdata/impl/fail_extend_impl_forall.carbon:
```diff
@@ -199,10 +201,10 @@ class C {
-// CHECK:STDOUT:
 // CHECK:STDOUT: specific @impl(%T.loc20_23.2) {}
 // CHECK:STDOUT:
+// CHECK:STDOUT: specific @GenericInterface(<unexpected>.inst68.loc20_54) {}
```
And here

danakj commented on toolchain/check/testdata/impl/fail_extend_partially_defined_interface.carbon:
```diff
@@ -91,13 +91,15 @@ interface I {
 // CHECK:STDOUT:
 // CHECK:STDOUT: specific @C(constants.%Self) {}
 // CHECK:STDOUT:
+// CHECK:STDOUT: specific @C(<unexpected>.inst21) {}
```
And here

danakj commented on toolchain/check/testdata/impl/min_prelude/forward_decls.carbon:
```diff
@@ -2034,17 +2034,19 @@ interface I {
 // CHECK:STDOUT: specific @impl(%Self) {}
 // CHECK:STDOUT:
-// CHECK:STDOUT: specific @C(@F.%Self) {}
+// CHECK:STDOUT: specific @C(<unexpected>.inst47.loc20_10) {}
```
And here

danakj commented on toolchain/check/testdata/impl/min_prelude/forward_decls.carbon:
```diff
@@ -2194,11 +2200,9 @@ interface I {
 // CHECK:STDOUT: specific @impl(%Self) {}
 // CHECK:STDOUT:
-// CHECK:STDOUT: specific @C(@F.%Self) {}
+// CHECK:STDOUT: specific @C(<unexpected>.inst71.loc21_10) {}
```
And here

danakj commented on toolchain/check/testdata/interface/no_prelude/fail_add_member_outside_definition.carbon:
```diff
@@ -126,9 +126,11 @@ interface Outer {
 // CHECK:STDOUT: }
 // CHECK:STDOUT:
-// CHECK:STDOUT: specific @F.2(constants.%Self.277, constants.%Self.b60) {}
+// CHECK:STDOUT: specific @Inner(<unexpected>.inst29) {}
```
And here

zygoloid committed:
Apply suggestions from code review

Co-authored-by: Dana Jansens <danakj@orodu.net>

zygoloid committed:
Make AddGenericConstantInstToEvalBlock consistently operate on the
current generic.

Previously it did a mixture of operating on the current generic and
operating on the generic passed as an argument.

zygoloid committed:
Stop passing around the constants_in_generic map so much.

We're already assuming we're operating on the current generic; we can
grab its map directly.

zygoloid committed:
Merge branch 'trunk' into toolchain-generics-on-the-fly-2

zygoloid committed:
Review comment. Clang-format.

zygoloid committed:
Exclude specifics that were never resolved from formatted SemIR.

Including these adds clutter to the output, and their arguments might
not have instruction names.

zygoloid committed:
Merge branch 'trunk' into toolchain-generics-on-the-fly-2

danakj commented:
Oh, apologies I didn't review this today, just saw it was ready. Will review tomorrow.

danakj commented on toolchain/check/generic.cpp:
```diff
@@ -61,60 +63,48 @@ static auto AddGenericConstantInstToEvalBlock(
-      }
+            context().constant_values().GetInstId(const_id))) {
+      inst_id = result.value();
       CARBON_CHECK(inst_id.has_value());
```
Does this CHECK still make sense here? It's checking that the insts in the map have values now? Maybe that's fine.. but just checking. To me it feels like maybe it's in the wrong spot now, if that's its job now, and if we wanted to check this we could do it when making the map?

zygoloid replied:
Yeah, good point. I don't think this is useful, removed.

danakj commented on toolchain/check/generic.cpp:
```diff
@@ -61,60 +63,48 @@ static auto AddGenericConstantInstToEvalBlock(
+  }
 
   // Check for instructions for which we already have a mapping into the eval
   // block, and substitute them for the instructions in the eval block.
```
```suggestion
  // block, and substitute them with the instructions in the eval block.
```
Would this be correct? We look for an inst in the eval block, and replace `inst_id` with the inst from the eval block, right?

The way this reads right now, it sounds like the opposite to me. I noticed this while reading and thinking through your PR comment below.

danakj commented on toolchain/check/generic.cpp:
```diff
@@ -61,60 +63,48 @@ static auto AddGenericConstantInstToEvalBlock(
       return true;
     }
 
     // If this instruction is in the map, return the known result.
```
```suggestion
    // If this constant value has a defining instruction in the eval block, replace the
    // instruction in the body of the generic with the one from the eval block.
```
Is this right then? It would be helpful I think to explain the relationship here in terms of the instructions instead of just the fact we use a map.

danakj commented on toolchain/check/generic.cpp:
```diff
@@ -228,70 +200,66 @@ class RebuildTemplateActionInEvalBlockCallbacks final
 }
 
 // Adds instructions to compute the substituted value of `inst_id` in each
 // specific into the eval block for the generic, which is the current
 // instruction block. Returns a symbolic constant instruction ID that refers to
```
"which is the current instruction block" is no longer correct, I think?

zygoloid replied:
Done.

danakj commented on toolchain/check/generic.cpp:
```diff
@@ -228,70 +200,66 @@ class RebuildTemplateActionInEvalBlockCallbacks final
 // specific into the eval block for the generic, which is the current
 // instruction block. Returns a symbolic type ID that refers to the substituted
```
"which is the current instruction block" is no longer correct, I think?

zygoloid replied:
Done.

danakj commented on toolchain/check/testdata/if_expr/fail_not_in_function.carbon:
```diff
@@ -154,7 +154,7 @@ fn F() {
 // CHECK:STDOUT:
 // CHECK:STDOUT: class @B {
-// CHECK:STDOUT:   %C.ref: %C.type = name_ref C, <unexpected>.inst21.loc4_24 [concrete = constants.%C.generic]
+// CHECK:STDOUT:   %C.ref: %C.type = name_ref C, <unexpected>.inst23.loc4_24 [concrete = constants.%C.generic]
```
There's still some `<unexpected>` left, but in a very different context than the specifics one

zygoloid replied:
All that changed in this PR here was `inst21` -> `inst23`; the `<unexpected>` is pre-existing.

danakj commented on toolchain/check/testdata/alias/fail_control_flow.carbon:
```diff
@@ -79,7 +79,7 @@ fn F() {
 // CHECK:STDOUT:
 // CHECK:STDOUT: class @B {
-// CHECK:STDOUT:   %C.ref: %C.type = name_ref C, <unexpected>.inst32.loc4_24 [concrete = constants.%C.generic]
+// CHECK:STDOUT:   %C.ref: %C.type = name_ref C, <unexpected>.inst34.loc4_24 [concrete = constants.%C.generic]
```
Here also, another `name_ref``.

zygoloid replied:
Also pre-existing.

zygoloid committed:
Fix direction of comment.

Co-authored-by: Dana Jansens <danakj@orodu.net>

zygoloid committed:
Improve comment.

Co-authored-by: Dana Jansens <danakj@orodu.net>

zygoloid committed:
Changes for some review comments.

zygoloid committed:
Add an accessor on `Specific` to determine if it was never resolved.

danakj reviewed (approved):
LGTM

zygoloid committed:
Merge branch 'trunk' into toolchain-generics-on-the-fly-2

zygoloid committed:
Merge branch 'trunk' into toolchain-generics-on-the-fly-2

zygoloid committed:
Merge branch 'trunk' into toolchain-generics-on-the-fly-2

zygoloid committed:
Format after merg

zygoloid committed:
Merge branch 'trunk' into toolchain-generics-on-the-fly-2

zygoloid committed:
Merge branch 'trunk' into toolchain-generics-on-the-fly-2

danakj reviewed (commented):
Want to confirm this isn't waiting for anything from me right?

danakj commented:
> [<img alt="@zygoloid" width="20" height="20" src="https://avatars.githubusercontent.com/u/232783?s=40&amp;v=4">](/zygoloid) [zygoloid](/zygoloid) enabled auto-merge [May 1, 2025 15:06](#event-17491302567)

>

danakj commented:
> > [<img alt="@zygoloid" width="20" height="20" src="https://avatars.githubusercontent.com/u/232783?s=40&amp;v=4">](/zygoloid) [zygoloid](/zygoloid) enabled auto-merge [May 1, 2025 15:06](#event-17491302567)
> 
> >

I guess this says it's not. At which point I clicked the wrong button.. sorry 😓 should be in queue again.
